### ISSUE TITLE
Make Netty 5 a general-purpose event framework

### ISSUE THREAD

**Issue by almson** [netty/netty#12317]
Netty bills itself as an "asynchronous event-driven network application framework." Here is what I would suggest to help drop the "network" part and make it into an "asynchronous event-driven application framework."

  - Make EmbeddedChannel (or as I'd call it EmbeddedPipeline or maybe Pipeline, see below) a first-class citizen
  - Clear distinction between Channel (external connection) and Pipeline (internal bus) in naming
    - Pipeline instead of Channel+Pipeline. PipelineHandler instead of ChannelHandler. PipelineContext.
  - Unified interface for messages going in both directions
  	- unified handler method names (eg, handleInbound, handleOutbound instead of fireChannelRead and write. No userEventTriggered.)
  	- channel commands (not pipeline messages) to read, close, flush etc should be distinct, and either put into a ChannelCommandInterceptor with names like interceptChannelReadCommand, interceptChannelWriteCommand, or else simply made into messages that travel the pipeline
  	- promises in both directions (and on all commands including flush)
  - Unified exception handling
  	- Currently, three (afaik) paths for exceptions in handlerAdded, channelRead, write
  	- All unhandled exceptions hit `exceptionCaught(PipelineContext, PipelineDirection, Throwable)`
  	- read/write return promises, handlerAdded exception bubbles up to caller
  - Promise should allow attaching callbacks to callbacks (ie, `then`/`apply` and friends instead of `addListener`)
    and 
  - Handlers should have consistent names (ie, base class last)
    - DuplexPipelineHandler instead of ChannelDuplexHandler (and it should be an interface)
    - AbstractInboundPipelineHandler instead of ChannelInboundHandlerAdapter
    - AbstractDuplexPipelineHandler
  - The ChannelInitializer mechanism is weird. Why can't we pass a Pipeline or a list of handlers to Bootstrap?

**Comment by franz1981** [comment]
I'm not sure about it: right now channel operations are built around the idea of read/write(and flush) events (read) and actions (write/flush) and these are the sole event(s) available - they won't look in any form "generic" but the leak the idea that something must be read in and written to.
A generic event driven framework should be able to define and handle generic events other then these ones.
There are no ways to communicate across event loop(s) by using an event bus, but with just a very low level EventLoop API (that looks more like an enhanced scheduler-executor).

In short: Netty probably is leaking (with good reasons) the concept of being built to communicate through a network medium, while others (thinking about Vert-x) already provide higher-level abstraction(s) to hide this detail while providing additional features to be more generic and concise while processing "events".

**Comment by almson** [comment]
A little background of where I'm coming from. I've built an application on Netty that started as a simple, "accept connection, receive data, process it sequentially" and this application has grown in flexibility and complexity to where I'm creating a (reconfigurable) hydra with heads, tails, and many EmbeddedChannels. Netty works perfectly well for this, but I was frustrated and confused by the somewhat tedious interface.

I guess I'm not saying I want Netty to be a super-general-purpose framework. But that it's a bit too geared to *simple* pipelines and ends up more confusing as a result.

**Comment by chrisvest** [comment]
I'm with Franz on this. Netty is first and foremost about network IO, and there are particular semantics and behaviours there we need to be able to express with clarity.
Adding convenience methods, like the combinators we've already added to `Future` in Netty 5, are better discussed on a case-by-case basis.
With that in mind, I'm going to close this issue.

**Comment by almson** [comment]
I think the headline came out a bit sensational.

I'm just proposing some API renaming and refactoring to ease the learning curve. Please consider the individual proposals.
